File: cpp/submod/ProbabilisticSetCover.cpp
#include<iostream>
#include<vector>
#include<string>
#include<algorithm>
#include<cmath>
#include<set>
#include<iterator>
#include<map>
#include"ProbabilisticSetCover.h"

ProbabilisticSetCover::ProbabilisticSetCover(ll n_, std::vector<std::vector<float>> const &groundSetConceptProbabilities_, int numConcepts_, std::vector<float> const& conceptWeights_): n(n_), groundSetConceptProbabilities(groundSetConceptProbabilities_), numConcepts(numConcepts_), conceptWeights(conceptWeights_)  {
	probOfConceptsCoveredByX = std::vector<double>(numConcepts, 1);
}

// ProbabilisticSetCover* ProbabilisticSetCover::clone() {
// 	return NULL;
// }

double ProbabilisticSetCover::evaluate(std::unordered_set<ll> const &X) {
	double result=0;

	if(X.size()==0) {
		return 0;
	}

	for(int i=0; i < numConcepts; i++) {
		double product = 1;
		for(auto elem: X) {
			product *= (1-groundSetConceptProbabilities[elem][i]);
		}
		result += conceptWeights[i]*(1-product);
	}
	return result;
}

double ProbabilisticSetCover::evaluateWithMemoization(std::unordered_set<ll> const &X) { 
	double result = 0;
	if(X.size()==0) {
		return 0;
	}
	for(int i=0; i < numConcepts; i++) {
		result += conceptWeights[i]*(1-probOfConceptsCoveredByX[i]);
	}
	return result;
}


double ProbabilisticSetCover::marginalGain(std::unordered_set<ll> const &X, ll item) {
	double gain = 0;
	if (X.find(item)!=X.end()) {
		return 0;
	}
	
	for(int i=0; i<numConcepts; i++) {
		double oldConceptProd = 1;
		for(auto elem: X) {
			oldConceptProd *= (1-groundSetConceptProbabilities[elem][i]);
		}
		gain += conceptWeights[i]*oldConceptProd*groundSetConceptProbabilities[item][i];
	}
	return gain;
}


double ProbabilisticSetCover::marginalGainWithMemoization(std::unordered_set<ll> const &X, ll item, bool enableChecks) {
	double gain = 0;
	if (enableChecks && X.find(item)!=X.end()) {
		return 0;
	}
	for(int i=0; i<numConcepts; i++) {
		gain += conceptWeights[i]*probOfConceptsCoveredByX[i]*groundSetConceptProbabilities[item][i];
	}
	return gain;
}

void ProbabilisticSetCover::updateMemoization(std::unordered_set<ll> const &X, ll item) {
	if (X.find(item)!=X.end()) {
		return;
	}
	for(int i=0; i<numConcepts; i++) {
		probOfConceptsCoveredByX[i] *= (1-groundSetConceptProbabilities[item][i]);
	}
}

std::unordered_set<ll> ProbabilisticSetCover::getEffectiveGroundSet() {
	std::unordered_set<ll> effectiveGroundSet;
	effectiveGroundSet.reserve(n);
	for (ll i = 0; i < n; ++i){
		effectiveGroundSet.insert(i); 
	}
	return effectiveGroundSet;
}


// std::vector<std::pair<ll, double>> ProbabilisticSetCover::maximize(std::string optimizer,ll budget, bool stopIfZeroGain=false, bool stopIfNegativeGain=false, float epsilon = 0.1, bool verbose=false, bool showProgress=true) {
// 	if(optimizer == "NaiveGreedy") {
// 		return NaiveGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, verbose, showProgress);
// 	} else if(optimizer == "LazyGreedy") {
//         return LazyGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, verbose, showProgress);
// 	} else if(optimizer == "StochasticGreedy") {
//         return StochasticGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon, verbose, showProgress);
// 	} else if(optimizer == "LazierThanLazyGreedy") {
//         return LazierThanLazyGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon, verbose, showProgress);
// 	} else {
// 		throw "Error: Invalid Optimizer";
// 	}
// }

void ProbabilisticSetCover::clearMemoization() {
	for(int i=0; i<numConcepts; i++) {
		probOfConceptsCoveredByX[i] = 1;
	}
}

void ProbabilisticSetCover::setMemoization(std::unordered_set<ll> const &X) 
{
    clearMemoization();
    std::unordered_set<ll> temp;
    for (auto elem: X)
	{	
		updateMemoization(temp, elem);
		temp.insert(elem);	
	}
}

File: pytorch/submod/ProbabilisticSetCover.py
import torch
from typing import List, Set, Tuple
from ..SetFunction import SetFunction

class ProbabilisticSetCover(SetFunction):
    def __init__(self, n: int, ground_set_concept_probabilities: List[List[float]], num_concepts: int, concept_weights: List[float] = None):
        super(SetFunction, self).__init__()
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.n = n
        self.ground_set_concept_probabilities = ground_set_concept_probabilities
        self.num_concepts = num_concepts
        self.concept_weights = concept_weights

        if self.concept_weights is None:
            self.concept_weights = [1.0] * num_concepts
        else:
            self.concept_weights = torch.tensor(concept_weights, dtype=torch.float32).to(device)
        self.prob_of_concepts_covered_by_X = num_concepts

    def evaluate(self, X: Set[int]) -> float:
        result = 0
        if not X:
            return result

        for i in range(self.num_concepts):
            product = 1
            for elem in X:
                product *= (1 - self.ground_set_concept_probabilities[elem][i])
            result += self.concept_weights[i] * (1 - product)

        return result

    def evaluate_with_memoization(self, X: Set[int]) -> float:
        result = 0
        if not X:
            return result

        for i in range(self.num_concepts):
            result += self.concept_weights[i] * (1 - self.prob_of_concepts_covered_by_X[i])

        return result

    def marginal_gain(self, X: Set[int], item: int) -> float:
        gain = 0
        if item in X:
            return gain

        for i in range(self.num_concepts):
            old_concept_prod = 1
            for elem in X:
                old_concept_prod *= (1 - self.ground_set_concept_probabilities[elem][i])
            gain += self.concept_weights[i] * old_concept_prod * self.ground_set_concept_probabilities[item][i]
        return gain

    def marginal_gain_with_memoization(self, X: Set[int], item: int, enable_checks: bool = True) -> float:
        gain = 0
        if enable_checks and item in X:
            return gain
        for i in range(self.num_concepts):
            gain += self.concept_weights[i] * self.prob_of_concepts_covered_by_X[i] * self.ground_set_concept_probabilities[item][i]
        return gain

    def update_memoization(self, X: Set[int], item: int):
        if item in X:
            return

        for i in range(self.num_concepts):
            self.prob_of_concepts_covered_by_X[i] *= (1 - self.ground_set_concept_probabilities[item][i])

    def get_effective_ground_set(self) -> Set[int]:
        return set(range(self.n))

    def clear_memoization(self):
        self.prob_of_concepts_covered_by_X = torch.ones(self.num_concepts, dtype=torch.double)

    def set_memoization(self, X: Set[int]):
        self.clear_memoization()
        temp = set()
        for elem in X:
            self.update_memoization(temp, elem)
            temp.add(elem)
